tests/api_workflow/mocked_api_workflow_client.py

```diff
@@ -298,7 +336,32 @@ def upload_file_with_signed_url(
         res = Response()
         return res
 
+    def _get_csv_reader_from_read_url(self, read_url: str):
+        n_rows: int = self.n_embedding_rows_on_server
+        n_dims: int = self.n_dims_embeddings_on_server
+
+        rows_csv = [['filenames'] + [f'embeddings_{i}' for i in range(n_dims)] + ['labels']]
+        for i in range(n_rows):
+            row = [f'{self.embeddings_filename_base}_{i}.jpg']
+            for _ in range(n_dims):
+                row.append(np.random.uniform(0, 1))
+            row.append(i)
+            rows_csv.append(row)
+
+        # save the csv rows in a temporary in-memory string file
+        # using a csv writer and then read them as bytes
+        f = tempfile.SpooledTemporaryFile(mode="rw")
+        writer = csv.writer(f)
+        writer.writerows(rows_csv)
+        f.seek(0)
+        buffer = io.StringIO(f.read())
```

Comment:
this seems like we are loading the CSV 3-4 times into memory?!
- first when building `rows_csv`
- second when we write it to the temporary file `f`
- third when we read it via `StringIO(f.read())` in to the buffer
- forth? when we call `csv.reader(`

That seems a bit of an overkill, no? Is not possible to directly make a buffer from a dict struct?

Reply:
This is for mocking, so performance for the files with <100 rows does not matter. Furthermore there is the problem that you cannot directly write into a string buffer and read from it byte-wise. Thus I would keep this working implementation.